JsonNull return false when comparing against null

henrikse55 opened the pull request:
I'ved modified JsonNulls comparison methods to return false when
compared against null, the matching test have also been updated
in change

Fix #820

henrikse55 committed:
JsonNull return false when comparing against null

I'ved modified JsonNulls comparison methods to return false when
compared against null, the matching test have also been updated
in change

Fix #820

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=842) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=842) before we can accept your contribution.<br/><br/>:x: henrikse55 [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=842)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=842) it.</sub>

henrikse55 commented:
I do not believe the current tests which are failing are related to the changes I made as it network timeouts.

I also have a couple of questions as this is my first PR:

As it's just a timeout, should the failed tests be rerun?
When I think it's ready for review, should I tag the issuer?

@ahsonkhan

khellang commented on src/libraries/System.Text.Json/src/System/Text/Json/Node/JsonNull.cs:
```diff
@@ -43,23 +43,34 @@ public sealed class JsonNull : JsonNode, IEquatable<JsonNull>
         /// <param name="other">The JSON null to compare against.</param>
         /// <returns><see langword="true"/></returns>
-        public bool Equals(JsonNull other) => true;
+        public bool Equals(JsonNull other) => !(other is null);
```
What's the policy on `other is object`? 🤔

henrikse55 replied:
After looking at the code again I noticed I missed some change to `Equals(object obj)` method

https://github.com/dotnet/runtime/blob/f2a9f007afac5a8a9f48c0edb2603b5830ad7fcc/src/libraries/System.Text.Json/src/System/Text/Json/Node/JsonNull.cs#L33
Should have been
`public override bool Equals(object obj) => obj is JsonNull jsonNull && Equals(jsonNull);`

if this didn't cover your question would you mind rephrasing what you mean?

khellang replied:
I'm just wondering what the stance is on the `x is object` syntax is. The aspnet folks don't seem to love it, but it's equivalent to `!(x is null)`, see https://sharplab.io/#v2:CYLg1APgAgzABFATHAwgWAFAG9Nzw+AIwHtiAbOAMVIApjCArAUwGMAXOADwEpd8cM+IQgDscAIQ1OcAJYBnOADsArmTLcA3HD54Avju2D8sOCXJwAQgEMATnUasOPQ0IHDjY6fLj1m7DQa6hrpAA===

henrikse55 replied:
I see, I did `!(x is null)` as that would follow how the other Json methods do it.

JsonString:

https://github.com/dotnet/runtime/blob/f2a9f007afac5a8a9f48c0edb2603b5830ad7fcc/src/libraries/System.Text.Json/src/System/Text/Json/Node/JsonString.cs#L168

JsonBoolean:
https://github.com/dotnet/runtime/blob/f2a9f007afac5a8a9f48c0edb2603b5830ad7fcc/src/libraries/System.Text.Json/src/System/Text/Json/Node/JsonBoolean.cs#L57

ahsonkhan replied:
Given they are quite similar (identical?), I would follow what other `JsonNode` types do, to remain internally consistent.

> After looking at the code again I noticed I missed some change to `Equals(object obj)` method

https://github.com/dotnet/runtime/blob/f2a9f007afac5a8a9f48c0edb2603b5830ad7fcc/src/libraries/System.Text.Json/src/System/Text/Json/Node/JsonNull.cs#L33
> Should have been
`public override bool Equals(object obj) => obj is JsonNull jsonNull && Equals(jsonNull);`

Agreed. Please address that and add the appropriate test for that case :)

ahsonkhan commented:
> As it's just a timeout, should the failed tests be rerun?

Yes. However, what I tend to do (and encourage) is that I do a cursory search of the test failure/exception in the repo to see if an issue already exists tracking the intermittent failure, and if not, then file one, so that it gets tracked and resolved. Here are some examples: https://github.com/dotnet/corefx/issues/42583, https://github.com/dotnet/corefx/issues/42077, https://github.com/dotnet/corefx/issues/41881

> When I think it's ready for review, should I tag the issuer?

Generally, you'd tag the area owners to help review the PR:
https://github.com/dotnet/runtime/blob/master/docs/area-owners.md

I am happy to look it over and provide feedback.

ahsonkhan commented on src/libraries/System.Text.Json/src/System/Text/Json/Node/JsonNull.cs:
```diff
@@ -43,23 +43,34 @@ public sealed class JsonNull : JsonNode, IEquatable<JsonNull>
         /// </summary>
         /// <param name="other">The JSON null to compare against.</param>
         /// <returns><see langword="true"/></returns>
```
We should update the xml docs accordingly. For example:
https://github.com/dotnet/runtime/blob/3fb2aba3df7032c1267e95ca34a67d7852ed685f/src/libraries/System.Text.Json/src/System/Text/Json/Node/JsonBoolean.cs#L53-L56

ahsonkhan commented on src/libraries/System.Text.Json/tests/JsonNullTests.cs:
```diff
@@ -46,16 +46,16 @@ public static void TestEquals()
             Assert.False(jsonNull1.Equals(new Exception()));
 
             // Null comparisons behave this way because of implicit conversion from null to JsonNull:
```
Do we have an implicit conversion from `null` to `JsonNull`? How?

Looks like we should remove this comment. @joperezr, @bartonjs - do you know what this comment is trying to say?

bartonjs replied:
JsonNode defines many implicit conversions, the one from string can produce a JsonNull.

```C#
string s = null;
JsonNode sNode = s;
// sNode is JsonNull.Instance
```

If the calling code was

```C#
JsonNode someNode = ...;
string s = null;

if (someNode == s)
{
    ...
}
```

But since we don't define `operator==` or IEquatable on JsonNode (only the sealed node types), I don't see where ambiguity comes in.

Maybe this is left over from an earlier, more flexible model.  But I agree it should be deleted.

ahsonkhan replied:
> JsonNode defines many implicit conversions, the one from string can produce a JsonNull.

Yes, of course. Thanks!

> But I agree it should be deleted.

Great.

ahsonkhan reviewed (approved):
Other than updating xml doc comments, looks good.

ahsonkhan commented:
> I do not believe the current tests which are failing are related to the changes I made as it network timeouts.

Right, the test failure in CI is unrelated: https://github.com/dotnet/corefx/issues/20069#issuecomment-566292587

henrikse55 committed:
Updated xml comments to reflect my changes

I have updated the xml comments to reflect the changes
that was made in JsonNull and removed left over comment
in JsonNullTests

henrikse55 commented:
I have updated the XML docs and this should be ready for a review
@ahsonkhan @steveharter 

The test failure looked unrelated to me and I've commented on #702

ahsonkhan commented on src/libraries/System.Text.Json/src/System/Text/Json/Node/JsonNull.cs:
```diff
@@ -42,24 +42,44 @@ public sealed class JsonNull : JsonNode, IEquatable<JsonNull>
-        /// <returns><see langword="true"/></returns>
-        public static bool operator ==(JsonNull left, JsonNull right) => true;
+        /// <returns>
+        ///    <see langword="true"/> if both instances matches,
```
```suggestion
        ///    <see langword="true"/> if both instances match,
```

ahsonkhan commented on src/libraries/System.Text.Json/src/System/Text/Json/Node/JsonNull.cs:
```diff
@@ -42,24 +42,44 @@ public sealed class JsonNull : JsonNode, IEquatable<JsonNull>
-        /// <returns><see langword="true"/></returns>
-        public bool Equals(JsonNull other) => true;
+        /// <returns>
+        ///    <see langword="true"/> if the <paramref name="other"/> is <see cref="JsonNull"/>,
```
This comment doesn't match the API. `other` is already a `JsonNull`.

```suggestion
        ///    <see langword="true"/> if <paramref name="other"/> is not null,
```

ahsonkhan commented:
> The test failure looked unrelated to me and I've commented on #702 

The failure is related to this: https://github.com/dotnet/runtime/pull/737#issuecomment-566427331

henrikse55 committed:
Applied suggested change

Applied the suggested change to correct xml documentation

Co-Authored-By: Ahson Khan <ahkha@microsoft.com>

henrikse55 committed:
Applied Suggested change

Applied the suggested change to correct xml documentation

Co-Authored-By: Ahson Khan <ahkha@microsoft.com>

ahsonkhan commented:
Thanks for fixing this @henrikse55 - let me know if you are interested in picking up other issues :)

https://github.com/dotnet/runtime/issues?q=is%3Aopen+is%3Aissue+label%3Aarea-System.Text.Json+label%3Aup-for-grabs
https://github.com/dotnet/corefx/issues?q=is%3Aopen+is%3Aissue+label%3Aarea-System.Text.Json+label%3Aup-for-grabs
